app/api/helpers/mail.py

```diff
@@ -35,40 +54,53 @@ def send_email(to, action, subject, html, attachments=None):
         }
 
         if not current_app.config['TESTING']:
-            if email_service == 'smtp':
-                smtp_encryption = get_settings()['smtp_encryption']
-                if smtp_encryption == 'tls':
-                    smtp_encryption = 'required'
-                elif smtp_encryption == 'ssl':
-                    smtp_encryption = 'ssl'
-                elif smtp_encryption == 'tls_optional':
-                    smtp_encryption = 'optional'
+            smtp_encryption = get_settings()['smtp_encryption']
+            if smtp_encryption == 'tls':
+                smtp_encryption = 'required'
+            elif smtp_encryption == 'ssl':
+                smtp_encryption = 'ssl'
+            elif smtp_encryption == 'tls_optional':
+                smtp_encryption = 'optional'
+            else:
+                smtp_encryption = 'none'
+
+            config = {
+                'host': get_settings()['smtp_host'],
+                'username': get_settings()['smtp_username'],
+                'password': get_settings()['smtp_password'],
+                'encryption': smtp_encryption,
+                'port': get_settings()['smtp_port'],
+            }
+            smtp_status = check_smtp_config(smtp_encryption)
+            if smtp_status:
+                if email_service == 'smtp':
+                    try:
+                        send_email_task.delay(payload=payload, headers=None, protocol='smtp', config=config)
+                    except Exception as e:
+                        logging.exception("Error while sending email-{}".format(str(e)))
                 else:
-                    smtp_encryption = 'none'
-
-                config = {
-                    'host': get_settings()['smtp_host'],
-                    'username': get_settings()['smtp_username'],
-                    'password': get_settings()['smtp_password'],
-                    'encryption': smtp_encryption,
-                    'port': get_settings()['smtp_port'],
-                }
-
-                from .tasks import send_mail_via_smtp_task
-                send_mail_via_smtp_task.delay(config, payload)
+                    try:
+                        key = get_settings()['sendgrid_key']
+                        headers = {
+                            "Authorization": ("Bearer " + key),
+                            "Content-Type": "application/json"
+                        }
+                        try:
+                            payload['fromname'] = email_from_name
+                            send_email_task.delay(payload=payload, headers=headers, protocol='sendgrid')
+
+                        except requests.exceptions.HTTPError as e:
```

Comment:
I'm confident this will never get called since you are using `.delay` above

Reply:
@iamareebjamal I think its better to handle the exceptions in the task functions itself. Will update.